apps/sms/services/js/moz_mobile_message/messaging_serivce.js

```diff
@@ -0,0 +1,179 @@
+      });
+
+      return Promise.all(requests).then((values) =>
+        values.map((value, idx) => ({
+          success: !event.target.error,
+          result: event.target.result,
+          code: event.target.error,
+          recipient: recipients[idx]
+        }));
+      );
+
+    },
+
+    /**
+     * Send mms by using sendMMS API shim.
+     * @param {*} opts Options for sending mms contains following properties:
+     * - recipients (string or array of string): recipients for this message.
+     * - subject (optional string): subject for this message.
+     * - content (array of SMIL slides): this is the content for the message (
+     *   see ConversationView for more information).
+     * - serviceId (optional long or string): the SIM that should be used for
+     *   sending this message. If this is not the current default configuration
+     *   for sending MMS, then we'll first switch the configuration to this
+     *   serviceId, and only then send the message. That means that the
+     *   "sending" event will come quite late in this case.
+     * @returns {Promise} DOMRequest promise with success result or error.
+     */
+    sendMMS: function mm_sendMMS(opts) {
+      var recipients = opts.recipients,
+          subject = opts.subject,
+          content = opts.content,
+          serviceId = opts.serviceId = sanitizeServiceId(opts.serviceId),
+
+      if (!Array.isArray(recipients)) {
+        recipients = [recipients];
+      }
+
+      var message = SMIL.generate(content);
+
+      var sendOpts = getSendOptionsFromServiceId(serviceId);
+
+      return this.client('sendMMS', {
+        receivers: recipients,
+        subject: subject,
+        smil: message.smil,
+        attachments: message.attachments
+      }, sendOpts).then(
+        (event) => event.target.result
+      ).catch(
+        (event) => event.target.error
+      );
+    },
+
+    /**
+     * Resend a message by send API shim with exsiting message information.
+     * @param {Message} a formatted message in case you happen to have one.
+     * @returns {Promise} DOMRequest promise with success result or error.
+     */
+    resendMessage: function mm_resendMessage(message) {
+      if (!message) {
+        throw new Error('Message to resend is not defined.');
+      }
+
+      var serviceId = Settings.getServiceIdByIccId(message.iccId);
+      var sendOpts = getSendOptionsFromServiceId(serviceId);
+      var request;
+
+      if (message.type === 'sms') {
+        request = this.client('send', message.receiver, message.body, sendOpts);
+      }
+      if (message.type === 'mms') {
+        request = this.client('sendMMS', {
+          receivers: message.receivers,
+          subject: message.subject,
+          smil: message.smil,
+          attachments: message.attachments
+        }, sendOpts);
+      }
+
+      return request.then((evt) => {
+        this.client('delete', message.id);
```

Comment:
Not sure we should make delete shim and request into promise chain since these 2 actions seems have no relationship. Wondering if we can move out delete shim from request.